env/ipp_aes_ctr_provider.cc

```diff
@@ -0,0 +1,204 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  Copyright (c) 2020 Intel Corporation
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#ifndef ROCKSDB_LITE
+
+#include "ipp_aes_ctr_provider.h"
+
+#include "util/string_util.h"
+
+#endif
+
+namespace ROCKSDB_NAMESPACE {
+
+#ifndef ROCKSDB_LITE
+
+Status IppAESCTRProvider::CreateProvider(
+    const std::string& id, std::shared_ptr<EncryptionProvider>* provider) {
+  if (id != kName()) {
+    return Status::NotSupported("Invalid provider ", id);
+  }
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+  IppAESCTRProvider* ptr_ipp = new IppAESCTRProvider();
+  provider->reset(ptr_ipp);
+  return Status::OK();
+#else
+  (void)id;
+  (void)provider;
+  return Status::Aborted("ipp-crypto library not found and requires SSE2+.");
+#endif  // IPPCP
+}
+
+Status IppAESCTRProvider::AddCipher(const std::string& /*descriptor*/,
+                                    const char* cipher, size_t len,
+                                    bool /*for_write*/) {
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+  // We currently don't support more than one encryption key
+  if (aes_ctx_ != nullptr) {
+    return Status::NotSupported("Multiple encryption keys not supported.");
+  }
+
+  // AES supports key sizes of only 16, 24, or 32 bytes
+  if (len != 16 && len != 24 && len != 32) {
+    return Status::NotSupported("Invalid key size in provider.");
+  }
+
+  size_t key_size = std::string(cipher).size();
+  if (key_size != len) {
+    return Status::NotSupported("Key size mismatch in provider.");
```

Comment:
I have removed this check entirely because I found no good way to do it without forbidding an embedded null character in the cipher which I don't want to do.